plugin/kubernetes/setup.go

```diff
@@ -39,8 +39,14 @@ func setup(c *caddy.Controller) error {
 	c.OnStartup(func() error {
 		go kubernetes.APIConn.Run()
 		if kubernetes.APIProxy != nil {
-			go kubernetes.APIProxy.Run()
+			kubernetes.APIProxy.Run()
 		}
+		synced := false
+		for synced == false {
```

Comment:
Please check https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L104

You can use
```
        if !cache.WaitForCacheSync(stopCh,
		kubernetes.APIConn.HasSynced,
	) {
		return fmt.Errorf("Timed out waiting for caches to sync")
	}
```
from package `k8s.io/client-go/tools/cache`

Basically this waits until the cache has synced.

Reply:
Interesting. I'll check tomorrow - might still merge this and then follow up.